HeterogeneousCore/AlpakaCore/python/ProcessAcceleratorAlpaka.py

```diff
@@ -45,26 +51,70 @@ class ProcessAcceleratorAlpaka(cms.ProcessAccelerator):
     ProcessAcceleratorCUDA) define.
     """
     def __init__(self):
-        super(ProcessAcceleratorAlpaka,self).__init__()
+        super(ProcessAcceleratorAlpaka, self).__init__()
         self._backend = None
+
     # User-facing interface
     def setBackend(self, backend):
         self._backend = backend
+
     # Framework-facing interface
     def moduleTypeResolver(self, accelerators):
         return ModuleTypeResolverAlpaka(accelerators, self._backend)
+
     def apply(self, process, accelerators):
-        if not hasattr(process, "AlpakaServiceSerialSync"):
+        # propagate the AlpakaService messages through the MessageLogger
+        if not hasattr(process.MessageLogger, "AlpakaService"):
+            process.MessageLogger.AlpakaService = cms.untracked.PSet()
+
+        # Check if the CPU backend is available
+        try:
+            if not "cpu" in accelerators:
+                raise False
             from HeterogeneousCore.AlpakaServices.AlpakaServiceSerialSync_cfi import AlpakaServiceSerialSync
-            process.add_(AlpakaServiceSerialSync)
-        if not hasattr(process, "AlpakaServiceCudaAsync"):
+        except:
+            # the CPU backend is not available, do not load the AlpakaServiceSerialSync
+            if hasattr(process, "AlpakaServiceSerialSync"):
+                del process.AlpakaServiceSerialSync
+        else:
+            # the CPU backend is available, load the AlpakaServiceSerialSync
+            if not hasattr(process, "AlpakaServiceSerialSync"):
+                process.add_(AlpakaServiceSerialSync)
+
+        # Check if CUDA is available, and if the system has at least one usable NVIDIA GPU
+        try:
+            if not "gpu-nvidia" in accelerators:
+                raise False
+            status = PlatformStatus(os.waitstatus_to_exitcode(os.system("cudaIsEnabled")))
+            if status != PlatformStatus.Success:
+                raise False
```

Comment:
Ehm, maybe ?
I mean that I do not know how the `accelerators` list gets populated, and if by the time `ProcessAcceleratorAlpaka.apply()` is called `gpu-nvidia` would already have been removed if there is no viable NVIDIA GPU in the system.

Reply:
The `apply()` function's `accelerators` argument is guaranteed to contain only those accelerator labels that are selected (via `process.options.accelerators`) and available (as dictated by `enabledLabels()). The availability of all accelerators is resolved first before the `apply()` function of any `ProcessAccelerator` is called. Therefore `ProcessAcceleratorAlpaka.apply()` should not see `gpu-nvidia` in case there system does not have a viable NVIDIA GPU.